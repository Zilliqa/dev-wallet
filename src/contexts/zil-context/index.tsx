/**
 * Copyright (c) 2018 - present Zilliqa Research Pte. Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';
import {
  getAddressFromPrivateKey,
  getPubKeyFromPrivateKey,
  fromBech32Address,
} from '@zilliqa-js/crypto';
import { Long, units, BN, validation } from '@zilliqa-js/util';
import { Transaction } from '@zilliqa-js/account';

import { bytes, Zilliqa } from '@zilliqa-js/zilliqa';
import { HTTPProvider, RPCMethod } from '@zilliqa-js/core';

export enum NETWORK {
  IsolatedServer = 'isolated_server',
  TestNet = 'testnet',
}

const initialState = {
  config: undefined as any,

  curNetwork: undefined as any,
  zilliqa: undefined as Zilliqa | undefined,
  version: undefined as number | undefined,
  provider: undefined as HTTPProvider | undefined,

  isAuth: undefined as boolean | undefined,
  address: undefined as string | undefined,
  publicKey: undefined as string | undefined,
  privateKey: undefined as string | undefined,
};
export const ZilContext = React.createContext(initialState);

export class ZilProvider extends React.Component {
  public readonly state = initialState;
  async componentDidMount() {
    const res = await fetch('config.json');
    const config = await res.json();
    console.log('config', config);
    this.setState({ config }, () => this.initState());
  }

  initState = (networkKey?: string) => {
    let curNetworkKey = networkKey || NETWORK.TestNet;

    const urlSearchParams = new URLSearchParams(window.location.search);
    const params = Object.fromEntries(urlSearchParams.entries());
    const networkParam = params['network'];
    const isValidNetwork = [NETWORK.IsolatedServer, NETWORK.TestNet].includes(
      networkParam as NETWORK
    );
    if (isValidNetwork) {
      curNetworkKey = networkParam;
    }

    if (networkParam === undefined || !isValidNetwork) {
      const urlSearchParams = new URLSearchParams(window.location.search);
      urlSearchParams.set('network', curNetworkKey);
      window.history.replaceState(
        null,
        '',
        `${window.location.pathname}?${urlSearchParams.toString()}`
      );
    }
    const { config } = this.state;

    const curNetwork =
      curNetworkKey === NETWORK.TestNet ? config[NETWORK.TestNet] : config[NETWORK.IsolatedServer];

    const provider = new HTTPProvider(curNetwork.nodeUrl);
    const zilliqa = new Zilliqa(curNetwork.nodeUrl, provider);
    const version = bytes.pack(curNetwork.chainId, curNetwork.msgVersion);

    this.setState({
      curNetwork: curNetwork,
      zilliqa: zilliqa,
      version: version,
      provider: provider,

      isAuth: undefined as boolean | undefined,
      address: undefined as string | undefined,
      publicKey: undefined as string | undefined,
      privateKey: undefined as string | undefined,
    });
  };

  public accessWallet = (privateKey: string) => {
    try {
      const address = getAddressFromPrivateKey(privateKey);
      const publicKey = getPubKeyFromPrivateKey(privateKey);
      const zilliqa = this.state.zilliqa as Zilliqa;
      zilliqa.wallet.addByPrivateKey(privateKey);

      this.setState({
        isAuth: true,
        privateKey,
        publicKey,
        address,
        zilliqa,
      });
    } catch (error) {
      this.setState({ isAuth: false });
    }
  };

  private getParams = async (toAddr, amountInZil) => {
    const zilliqa = this.state.zilliqa as Zilliqa;
    const version = this.state.version as number;
    const response = await zilliqa.blockchain.getMinimumGasPrice();
    const gasPrice: string = response.result || '';

    const amountInQa = units.toQa(amountInZil, units.Units.Zil);
    return {
      toAddr,
      version: version,
      amount: amountInQa,
      gasPrice: new BN(gasPrice.toString()),
      gasLimit: Long.fromNumber(50),
    };
  };

  public send = async ({ args }): Promise<any> => {
    const { amount, toAddress } = args;
    const zilliqa = this.state.zilliqa as Zilliqa;
    const provider = this.state.provider as HTTPProvider;
    const tx = new Transaction(await this.getParams(toAddress, amount), provider);
    const signedTx = await zilliqa.wallet.sign(tx);
    const { txParams } = signedTx;
    // Send a transaction to the network
    const res = await provider.send(RPCMethod.CreateTransaction, txParams);
    if (res.error !== undefined) throw new Error(res.error.message);
    return res.result ? res.result.TranID : undefined;
  };

  public getBalance = async (): Promise<string> => {
    const zilliqa = this.state.zilliqa as Zilliqa;
    const address = this.state.address as string;

    if (typeof address !== 'string') {
      return '0';
    }
    const res = await zilliqa.blockchain.getBalance(address);
    if (res.error !== undefined) return '0';
    return res.result ? res.result.balance : '0';
  };

  public getMinGasPrice = async (): Promise<string> => {
    const zilliqa = this.state.zilliqa as Zilliqa;
    const res = await zilliqa.blockchain.getMinimumGasPrice();
    if (res.error !== undefined) throw new Error(res.error.message);
    return res.result ? res.result : '0';
  };

  public faucet = async ({ args, signal }): Promise<string | void> => {
    const { token, toAddress } = args;

    let address = toAddress;
    // address is either ByStr20 or bech32
    //
    // ByStr20: 20 byte hexadecimal string
    // e.g. 0x573EC96638C8bB1c386394602E1460634F02aDdA
    //
    // bech32: A bech32 with a human-readable prefix of zil
    // e.g. zil12ulvje3ceza3cwrrj3szu9rqvd8s9tw69c978p

    if (validation.isBech32(toAddress)) {
      address = fromBech32Address(toAddress);
    }

    const body = JSON.stringify({
      address,
      token,
    });
    const { curNetwork } = this.state;
    const res = await fetch(curNetwork.faucetUrl, {
      signal,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body,
    });
    if (!res.ok) {
      throw new Error('Failed to run faucet, you may have reached maximum request limit.');
    }
    const data = await res.json();
    return data ? data.txId : undefined;
  };

  public clearAuth = () => {
    const { curNetwork } = this.state;
    this.initState(curNetwork.name);
  };

  public switchNetwork = (key) => {
    const urlSearchParams = new URLSearchParams(window.location.search);
    urlSearchParams.set('network', key);
    window.history.replaceState(
      null,
      '',
      `${window.location.pathname}?${urlSearchParams.toString()}`
    );
    this.initState(key);
  };

  public render() {
    return (
      <ZilContext.Provider
        value={
          {
            ...this.state,
            accessWallet: this.accessWallet,
            getBalance: this.getBalance,
            getMinGasPrice: this.getMinGasPrice,
            send: this.send,
            faucet: this.faucet,
            clearAuth: this.clearAuth,
            switchNetwork: this.switchNetwork,
          } as any
        }
      >
        {this.props.children}
      </ZilContext.Provider>
    );
  }
}
