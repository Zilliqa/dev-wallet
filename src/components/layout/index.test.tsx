/**
 * This file is part of nucleus-wallet.
 * Copyright (c) 2018 - present Zilliqa Research Pte. Ltd.
 *
 * nucleus-wallet is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * nucleus-wallet is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * nucleus-wallet.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render } from '@testing-library/react';
import Layout from '.';
import { MemoryRouter } from 'react-router';
import { NETWORK } from '../../contexts/zil-context';

test('matches the snapshot', () => {
  const clearAuth = jest.fn();
  const switchNetwork = jest.fn();
  const isAuth = false;
  const curNetwork = {
    name: NETWORK.TestNet,
    chainId: 333,
    msgVersion: 1,
    nodeUrl: 'https://dev-api.zilliqa.com',
    faucetUrl: 'https://nucleus-server.zilliqa.com/api/v1/run',
    explorerUrl: 'devex.zilliqa.com',
  };
  const zilContext = { curNetwork, isAuth, clearAuth, switchNetwork };

  const { container } = render(
    <MemoryRouter>
      <Layout zilContext={zilContext}>
        <div data-testid="children" />
      </Layout>
    </MemoryRouter>
  );
  expect(container).toMatchSnapshot();
});
