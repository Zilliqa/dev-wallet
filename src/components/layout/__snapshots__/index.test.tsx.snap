// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <div>
    <div>
      <nav
        class="navbar navbar-expand-sm navbar-dark bg-faded fixed-top"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Nucleus Wallet
          <small
            class="release-text"
          >
            beta
          </small>
        </a>
        <button
          aria-label="toggler"
          class="navbar-toggler"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
        >
          <ul
            class="ml-auto navbar-nav"
          >
            <li
              class="sidebar-link nav-item"
            >
              <a
                class="nav-link"
                href="/home"
              >
                Home
              </a>
            </li>
            <li
              class="sidebar-link nav-item"
            >
              <a
                class="nav-link"
                href="/generate"
              >
                Create New Wallet
              </a>
            </li>
            <li
              class="sidebar-link nav-item"
            >
              <a
                class="nav-link"
                href="/send"
              >
                Send ZIL
              </a>
            </li>
            <li
              class="sidebar-link nav-item"
            >
              <a
                class="nav-link"
                href="/faucet"
              >
                ZIL Faucet
              </a>
            </li>
            <span
              class="nav-link"
            >
              <span
                class="network"
              >
                Dev Testnet
              </span>
            </span>
          </ul>
        </div>
      </nav>
    </div>
    <div
      class="layout"
    >
      <div
        class="sidebar"
      >
        <div
          class="sidebar-background"
        >
          <div
            class="sidebar-wrapper"
          >
            <ul
              class="sidebar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link "
                  href="/home"
                >
                  <span
                    class="sidebar-icon pr-2"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 576 512"
                      width="1em"
                    >
                      <path
                        d="M488 312.7V456c0 13.3-10.7 24-24 24H348c-6.6 0-12-5.4-12-12V356c0-6.6-5.4-12-12-12h-72c-6.6 0-12 5.4-12 12v112c0 6.6-5.4 12-12 12H112c-13.3 0-24-10.7-24-24V312.7c0-3.6 1.6-7 4.4-9.3l188-154.8c4.4-3.6 10.8-3.6 15.3 0l188 154.8c2.7 2.3 4.3 5.7 4.3 9.3zm83.6-60.9L488 182.9V44.4c0-6.6-5.4-12-12-12h-56c-6.6 0-12 5.4-12 12V117l-89.5-73.7c-17.7-14.6-43.3-14.6-61 0L4.4 251.8c-5.1 4.2-5.8 11.8-1.6 16.9l25.5 31c4.2 5.1 11.8 5.8 16.9 1.6l235.2-193.7c4.4-3.6 10.8-3.6 15.3 0l235.2 193.7c5.1 4.2 12.7 3.5 16.9-1.6l25.5-31c4.2-5.2 3.4-12.7-1.7-16.9z"
                      />
                    </svg>
                  </span>
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link "
                  href="/generate"
                >
                  <span
                    class="sidebar-icon pr-2"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 448 512"
                      width="1em"
                    >
                      <path
                        d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                      />
                    </svg>
                  </span>
                  Create New Wallet
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link "
                  href="/send"
                >
                  <span
                    class="sidebar-icon pr-2"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
                      />
                    </svg>
                  </span>
                  Send ZIL
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link "
                  href="/faucet"
                >
                  <span
                    class="sidebar-icon pr-2"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 352 512"
                      width="1em"
                    >
                      <path
                        d="M205.22 22.09c-7.94-28.78-49.44-30.12-58.44 0C100.01 179.85 0 222.72 0 333.91 0 432.35 78.72 512 176 512s176-79.65 176-178.09c0-111.75-99.79-153.34-146.78-311.82zM176 448c-61.75 0-112-50.25-112-112 0-8.84 7.16-16 16-16s16 7.16 16 16c0 44.11 35.89 80 80 80 8.84 0 16 7.16 16 16s-7.16 16-16 16z"
                      />
                    </svg>
                  </span>
                  ZIL Faucet
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="content-section"
      >
        <div
          data-testid="children"
        />
        <footer
          class="footer"
          data-testid="footer"
        >
          <div
            class="text-center py-2"
          >
            <ul
              class="nav justify-content-center"
            >
              <li
                class="nav-item"
              >
                <a
                  aria-label="Zilliqa GitHub"
                  class="text-secondary nav-link"
                  href="https://github.com/Zilliqa"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 496 512"
                    width="1em"
                  >
                    <path
                      d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                    />
                  </svg>
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  aria-label="Zilliqa Gitter"
                  class="text-secondary nav-link"
                  href="https://gitter.im/Zilliqa"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 384 512"
                    width="1em"
                  >
                    <path
                      d="M66.4 322.5H16V0h50.4v322.5zM166.9 76.1h-50.4V512h50.4V76.1zm100.6 0h-50.4V512h50.4V76.1zM368 76h-50.4v247H368V76z"
                    />
                  </svg>
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  aria-label="Zilliqa Twitter"
                  class="text-secondary nav-link"
                  href="https://twitter.com/zilliqa"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
                    />
                  </svg>
                </a>
              </li>
            </ul>
            <span
              class="text-secondary copyright"
            >
              Copyright © 2019 Zilliqa Research Pte. Ltd.
            </span>
          </div>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
